[autoscaler] [azure] Fix Azure Autoscaling Failures

gramhagen opened the pull request:
- update node provider to support newer versions of azure-mgmt-resource sdk
 
azure autoscaler yaml updates: 
- use lastest vm image
- add pause after docker usermod for better robustness
- remove conda env in setup for docker based config
- update conda env to use python 3.8 versions

<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
ray up using azure yamls fails with various errors

## Related issue number

Closes #16091

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

I have manually tested using example_full.yaml, need to test other yamls (gpu, legacy, minimal)

gramhagen commented:
seeing more issues when setting up worker nodes... do not merge this yet

gramhagen commented:
I'm seeing a strange error on trying to deploy the worker nodes:
```
2021-06-24 01:19:02,577 INFO monitor.py:400 -- Starting monitor using ray installation: /anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/__init__.py
2021-06-24 01:19:02,577 INFO monitor.py:401 -- Ray version: 2.0.0.dev0
2021-06-24 01:19:02,577 INFO monitor.py:402 -- Ray commit: 3e2f6081459a4a6c79b33e633bffef1dc73be44b
2021-06-24 01:19:02,577 INFO monitor.py:403 -- Monitor started with command: ['/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py', '--logs-dir=/tmp/ray/session_2021-06-24_01-19-00_401345_10781/logs', '--redis-address=10.47.0.4:6379', '--logging-rotate-bytes=536870912', '--logging-rotate-backup-count=5', '--autoscaling-config=~/ray_bootstrap_config.yaml', '--redis-password=5241590000000000', '--monitor-ip=10.47.0.4']
2021-06-24 01:19:02,872 INFO monitor.py:143 -- Starting autoscaler metrics server on port 44217
2021-06-24 01:19:03,024 INFO monitor.py:151 -- Monitor: Started
2021-06-24 01:19:03,082 ERROR monitor.py:285 -- Error in monitor loop
Traceback (most recent call last):
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 316, in run
    self._initialize_autoscaler()
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/monitor.py", line 163, in _initialize_autoscaler
    prom_metrics=self.prom_metrics)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 95, in __init__
    self.reset(errors_fatal=True)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 573, in reset
    raise e
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/autoscaler.py", line 525, in reset
    self.config["cluster_name"])
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/providers.py", line 200, in _get_node_provider
    provider_cls = _get_node_provider_cls(provider_config)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/providers.py", line 176, in _get_node_provider_cls
    return importer(provider_config)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/providers.py", line 38, in _import_azure
    from ray.autoscaler._private.azure.node_provider import AzureNodeProvider
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/azure/node_provider.py", line 7, in <module>
    from azure.common.client_factory import get_client_from_cli_profile
ModuleNotFoundError: No module named 'azure.common'

==> /tmp/ray/session_latest/logs/monitor.out <==
```

this doesn't  make much sense because if I attach to the head node I can start python and run `from azure.common.client_factory import get_client_from_cli_profile` without a problem. 
However if I try to execute `ray monitor ray_bootstrap_config.yaml` I get the same error. 

@AmeerHajAli do you know of any reason modules would not be accessible when running the ray package that is installed using this wheel: `pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-2.0.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"`?

AmeerHajAli commented:
@DmitriGekhtman @ijrsvt , can you please help @gramhagen here?

gramhagen commented:
strangely, this doesn't occur when I use a previous wheel: https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-1.2.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl

philippeboyd commented:
@gramhagen I get the same error with Ray `1.4.0`

Head runs fine, but can't start a worker;  doing `from azure.common.client_factory import get_client_from_cli_profile`. directly on the head works like you said. I'm debugging at thee same time...

ijrsvt commented:
@gramhagen Which YAML are you using to launch the cluster?

philippeboyd commented:
@gramhagen it's essentially the same one as specified in the issue ([the azure example file](https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/azure/example-full.yaml)) except I don't install anything in the `setup_commands` since i already have everything in my docker image (ray `1.4.0`)

gramhagen commented:
this yaml works, but swapping out to use latest 2.0.0 wheel fails
```
# An unique identifier for the head node and workers of this cluster.
cluster_name: default

# The maximum number of workers nodes to launch in addition to the head
# node.
max_workers: 2

# The autoscaler will scale up the cluster faster with higher upscaling speed.
# E.g., if the task requires adding more nodes then autoscaler will gradually
# scale up the cluster in chunks of upscaling_speed*currently_running_nodes.
# This number should be > 0.
upscaling_speed: 1.0

# This executes all commands on all nodes in the docker container,
# and opens all the necessary ports to support the Ray cluster.
# Empty object means disabled.
docker: {}

# If a node is idle for this many minutes, it will be removed.
idle_timeout_minutes: 5

# Cloud-provider specific configuration.
provider:
    type: azure
    # https://azure.microsoft.com/en-us/global-infrastructure/locations
    location: westus2
    resource_group: ray-cluster
    # set subscription id otherwise the default from az cli will be used
    # subscription_id: 00000000-0000-0000-0000-000000000000

# How Ray will authenticate with newly launched nodes.
auth:
    ssh_user: ubuntu
    # you must specify paths to matching private and public key pair files
    # use `ssh-keygen -t rsa -b 4096` to generate a new ssh key pair
    ssh_private_key: ~/.ssh/id_rsa
    # changes to this should match what is specified in file_mounts
    ssh_public_key: ~/.ssh/id_rsa.pub

# More specific customization to node configurations can be made using the ARM template azure-vm-template.json file
# See documentation here: https://docs.microsoft.com/en-us/azure/templates/microsoft.compute/2019-03-01/virtualmachines
# Changes to the local file will be used during deployment of the head node, however worker nodes deployment occurs
# on the head node, so changes to the template must be included in the wheel file used in setup_commands section below

# Tell the autoscaler the allowed node types and the resources they provide.
# The key is the name of the node type, which is just for debugging purposes.
# The node config specifies the launch config and physical instance type.
available_node_types:
    ray.head.default:
        # The minimum number of worker nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 0
        # The maximum number of worker nodes of this type to launch.
        # This takes precedence over min_workers.
        max_workers: 0
        # The resources provided by this node type.
        resources: {"CPU": 2}
        # Provider-specific config, e.g. instance type.
        node_config:
            azure_arm_parameters:
                vmSize: Standard_D2s_v3
                # List images https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
                imagePublisher: microsoft-dsvm
                imageOffer: ubuntu-1804
                imageSku: 1804-gen2
                imageVersion: 21.01.21

    ray.worker.default:
        # The minimum number of nodes of this type to launch.
        # This number should be >= 0.
        min_workers: 1
        # The resources provided by this node type.
        resources: {"CPU": 2}
        # Provider-specific config, e.g. instance type.
        node_config:
            azure_arm_parameters:
                vmSize: Standard_D2s_v3
                # List images https://docs.microsoft.com/en-us/azure/virtual-machines/linux/cli-ps-findimage
                imagePublisher: microsoft-dsvm
                imageOffer: ubuntu-1804
                imageSku: 1804-gen2
                imageVersion: 21.01.21
                # optionally set priority to use Spot instances
                priority: Spot
                # set a maximum price for spot instances if desired
                # billingProfile:
                #     maxPrice: -1

# Specify the node type of the head node (as configured above).
head_node_type: ray.head.default

# Files or directories to copy to the head and worker nodes. The format is a
# dictionary from REMOTE_PATH: LOCAL_PATH, e.g.
file_mounts: {
#    "/path1/on/remote/machine": "/path1/on/local/machine",
#    "/path2/on/remote/machine": "/path2/on/local/machine",
     "~/.ssh/id_rsa.pub": "~/.ssh/id_rsa.pub"
}

# Files or directories to copy from the head node to the worker nodes. The format is a
# list of paths. The same path on the head node will be copied to the worker node.
# This behavior is a subset of the file_mounts behavior. In the vast majority of cases
# you should just use file_mounts. Only use this if you know what you're doing!
cluster_synced_files: []

# Whether changes to directories in file_mounts or cluster_synced_files in the head node
# should sync to the worker node continuously
file_mounts_sync_continuously: False

# Patterns for files to exclude when running rsync up or rsync down
rsync_exclude: []

# Pattern files to use for filtering out files when running rsync up or rsync down. The file is searched for
# in the source directory and recursively through all subdirectories. For example, if .gitignore is provided
# as a value, the behavior will match git's behavior for finding and using .gitignore files.
rsync_filter: []

# List of commands that will be run before `setup_commands`. If docker is
# enabled, these commands will run outside the container and before docker
# is setup.
initialization_commands:
    # get rid of annoying Ubuntu message
    - touch ~/.sudo_as_admin_successful

# List of shell commands to run to set up nodes.
setup_commands:
    # Note: if you're developing Ray, you probably want to create an AMI that
    # has your Ray repo pre-cloned. Then, you can replace the pip installs
    # below with a git checkout <your_sha> (and possibly a recompile).
    - (which conda && echo 'eval "$(conda shell.bash hook)"' >> ~/.bashrc) || true
    # - echo 'conda activate py37_pytorch' >> ~/.bashrc
    - (conda activate py37_tensorflow &> /dev/null && echo 'conda activate py37_tensorflow' >> ~/.bashrc) || true
    - which ray || pip install -U "ray[default] @ https://s3-us-west-2.amazonaws.com/ray-wheels/latest/ray-1.4.0.dev0-cp37-cp37m-manylinux2014_x86_64.whl"
    # Consider uncommenting these if you also want to run apt-get commands during setup
    # - sudo pkill -9 apt-get || true
    # - sudo pkill -9 dpkg || true
    # - sudo dpkg --configure -a

# Custom commands that will be run on the head node after common setup.
head_setup_commands:
    - pip install -U azure-cli-core==2.22.0 azure-mgmt-compute==14.0.0 azure-mgmt-msi==1.0.0 azure-mgmt-network==10.2.0 azure-mgmt-resource==13.0.0

# Custom commands that will be run on worker nodes after common setup.
worker_setup_commands: []

# Command to start ray on the head node. You don't need to change this.
head_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --head --port=6379 --object-manager-port=8076 --autoscaling-config=~/ray_bootstrap_config.yaml

# Command to start ray on worker nodes. You don't need to change this.
worker_start_ray_commands:
    - ray stop
    - ulimit -n 65536; ray start --address=$RAY_HEAD_IP:6379 --object-manager-port=8076

head_node: {}
worker_nodes: {}
```

ijrsvt commented:
Hmm @gramhagen I'm wondering if the autoscaler is starting without `conda` (i.e. that the conda environment is not being picked up by default).

gramhagen commented:
if so it's happening inside the ray code somehow (maybe ray or click is kicking off a new thread or not using the same shell?)
i can reproduce the issue by using a docker based yaml which doesn't rely on conda. after attaching to the head node and confirming azure-common in installed and exists in the site-packages folder if I run
`ray monitor ray_bootstrap_config.yaml` the error occurs again.
whereas just running `python -c "from azure.common.client_factory import get_client_from_cli_profile"` there's no error.
super weird... I wouldn't think ray would somehow inject something into the pythonpath which would cause a collision with azure?

I actually tried modifying `/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/autoscaler/_private/azure/node_provider.py` on the head node to add prints before the import and after. What is strange to me is the file seems to be loaded twice, the first time there is no error (both prints come through), the second time only the first print happens and then the import fails.

philippeboyd commented:
@gramhagen I went through the same kind of debugging process as you did. 

in the file `/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/monitor.py` I added the following import in first position just to test the import itself.
```
"""Autoscaler monitoring loop daemon."""
from azure.common.client_factory import get_client_from_cli_profile
# other imports...
```

and then simply callling `python /usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/monitor.py`

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/monitor.py", line 2, in <module>
    from azure.common.client_factory import get_client_from_cli_profile
ModuleNotFoundError: No module named 'azure.common'
```

I get the same error. I then noticed that there's also a package named `azure` in `/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/`

I'm assuming there's probably a conflict with the relative package `azure` in `/usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/`

When calling directly the file `python /usr/local/lib/python3.8/site-packages/ray/autoscaler/_private/azure/node_provider.py` I don't get the error...

gramhagen commented:
ooof, yeah that makes sense. curious why this only started popping up recently though. i thought it has been like that for a while.  or maybe the bug has also been around for a while =/

seems like we could just rename the folder or do some hacky modification to the sys.path

gramhagen commented:
i'm wondering now if that's why kubernetes is _kubernetes

gramhagen commented:
updated pr here to rename folder

ijrsvt commented:
@gramhagen Oh yes, that is why `kubernetes` was changed to `_kubernetes`. Python gets confused between the directory `azure` and the pip installed `azure` modules

ijrsvt commented:
Thanks @gramhagen and @philippeboyd for figuring this out =)

philippeboyd commented:
@gramhagen quick check on the modified files, I think you forgot azure's `node_provider.py` ?

```
from ray.autoscaler._private.azure.config import bootstrap_azure
```

philippeboyd commented:
@gramhagen don't forget the `setup.py` file also. Disregard if already done
![image](https://user-images.githubusercontent.com/3239656/123315171-6ae0e180-d4f9-11eb-9ca4-81c6d457ea0b.png)

gramhagen commented:
ack, thanks, i didn't think about search for actual folder paths

richardliaw commented:
Is this ready to be merged?

DmitriGekhtman commented:
There are some config test failures -- taking a look.

DmitriGekhtman commented on python/ray/autoscaler/azure/example-full-legacy.yaml:
```diff
@@ -129,13 +131,12 @@ setup_commands:
     # below with a git checkout <your_sha> (and possibly a recompile).
     # To run the nightly version of ray (as opposed to the latest), either use a rayproject docker image
     # that has the "nightly" (e.g. "rayproject/ray-ml:nightly-gpu") or uncomment the following line:
-    - echo 'eval "$(conda shell.bash hook)"' >> ~/.bashrc
```
Need an empty list here, rather than nil.
`setup_commands: []`

gramhagen replied:
thanks, updated!

DmitriGekhtman commented:
Besides minor typo mentioned above, lgtm!
@ijrsvt could you take a final look?

ijrsvt reviewed (approved):
LGTM, pending the addition Dmitri mentioned!

ijrsvt commented:
@gramhagen Just to double check, have you since tested:
[ ] gpu, 
[ ]legacy
[ ] minimal

gramhagen commented:
I need to build a new docker image and wheel to properly test these scenarios. Is there a way to have the ci system build an wheel that can be used?

DmitriGekhtman commented:
To test python changes, you can use setup commands to clone this branch and run `ray/python/ray/setup-dev.py` to replace the ray python libraries with the code in this branch, something like
```
setup_commands:
     - git clone --single-branch --branch fix_azure https://github.com/gramhagen/ray
     - pushd ray/python/ray && python setup-dev.py -y && popd
```
But, actually, given that there are changes to `setup.py` in this PR, I'm not sure if this would work. Definitely this shouldn't require rebuilding the Ray C++ source.
Any suggestions, @ijrsvt ?

ijrsvt commented:
Hmm, yeah @DmitriGekhtman. I'm not 100% sure what the best way to test this is. I think it is worth giving that a shot @gramhagen  (the `setup-dev.py` method).

gramhagen commented:
ok, testing now.
one thing that i noticed is the azure-mgmt-resource==13.0.0 package needs to be added to the docker ray-project/ray* docker images. should I add a separate pr for that?

DmitriGekhtman commented:
@gramhagen please feel free to add the module to `ray/docker/ray-deps/Dockerfile` in this PR

gramhagen commented:
@DmitriGekhtman, ok, I moved azure-mgmt-resource install to the Dockerfile (and removed it from the setup commands in the yamls). This does mean they won't work until the docker images are updated. Do you think that's safe or should I keep the install in the yaml too?

@ijrsvt I've tested the following yamls and they are working (using the setup-dev.py method and provided the package mentioned above is installed)

- [x] example-full-legacy
- [x] example-full
- [x] example-gpu-docker
- [ ] example-minimal

but when running example-minimal I get the following error
```
2021-07-06 22:52:56,796 ERROR services.py:1324 -- Failed to start the dashboard: Failed to start the dashboard, return code 0. The last 10 lines of /tmp/ray/session_2021-07-06_22-52-52_779842_16325/logs/dashboard.log:
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 215, in <module>
    log_dir=args.log_dir)
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 100, in __init__
    raise ex
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 91, in __init__
    build_dir = setup_static_dir()
  File "/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/dashboard.py", line 51, in setup_static_dir
    "&& npm run build)", build_dir)
FrontendNotFoundError: [Errno 2] Dashboard build directory not found. If installing from source, please follow the additional steps required to build the dashboard(cd python/ray/new_dashboard/client && npm install && npm ci && npm run build): '/anaconda/envs/py37_tensorflow/lib/python3.7/site-packages/ray/new_dashboard/client/build'
[2021-07-06 22:53:01,840 C 16325 16325] service_based_gcs_client.cc:72:  Check failed: GetGcsServerAddressFromRedis( redis_client_->GetPrimaryContext()->sync_context(), &current_gcs_server_address_, RayConfig::instance().gcs_service_connect_retries()) Failed to get gcs server address when init gcs client.
```

I've tried running the npm commands to build the dashboard, but that doesn't fix the gcs client timeout
any thoughts on how to resolve this?

richardliaw commented:
hmm, is it possible that there's a missing dependency?

or that it's only a transient error (does it happen multiple times)?

DmitriGekhtman commented:
> Do you think that's safe or should I keep the install in the yaml too?

Should be safe to have it just in the Dockerfile, except for defaults.yaml which doesn't use Docker -- for defaults, it needs to be in the yaml.

> any thoughts on how to resolve this?

There was some weirdness around the dashboard in Ray master yesterday. This might be resolved now?
example-minimal.yaml pulls defaults from defaults.yaml, which includes a setup command to install Ray master.
defaults.yaml (and thus example-minimal.yaml) are weird in that they don't use docker.

If you're still seeing the issue with example-minimal, could you post how you're modifying configs to test example-minimal?

gramhagen commented:
> > Do you think that's safe or should I keep the install in the yaml too?
> 
> Should be safe to have it just in the Dockerfile, except for defaults.yaml which doesn't use Docker -- for defaults, it needs to be in the yaml.

yeah, I kept the installation in defaults.yaml

> There was some weirdness around the dashboard in Ray master yesterday. This might be resolved now?
> example-minimal.yaml pulls defaults from defaults.yaml, which includes a setup command to install Ray master wheels.
> defaults.yaml (and thus example-minimal.yaml) are weird in that they don't use docker.
> 
> If you're still seeing the issue with example-minimal, could you post how you're modifying configs to test example-minimal?

that was it, I just merged in the latest changes and this went away, minimal now works too. so @richardliaw I think we're good with this one. Thanks for the help everyone!

ijrsvt commented on docker/ray-deps/Dockerfile:
```diff
@@ -17,7 +17,8 @@ RUN $HOME/anaconda3/bin/pip --no-cache-dir install $(basename $WHEEL_PATH)[all]
         "azure-mgmt-compute==14.0.0" \
         "azure-mgmt-msi==1.0.0" \
-        "azure-mgmt-network==10.2.0"; fi) \
+        "azure-mgmt-network==10.2.0"
```
```suggestion
        "azure-mgmt-network==10.2.0" \
```

gramhagen replied:
got it

ijrsvt commented on python/ray/autoscaler/azure/example-full-legacy.yaml:
```diff
@@ -118,24 +118,26 @@ rsync_filter:
 
 # Custom commands that will be run on the head node after common setup.
+# NOTE: rayproject/ray-ml:latest has azure packages bundled
 head_setup_commands:
```
```suggestion
head_setup_commands: []
```

Just to check, this is intentionally commented out still?

gramhagen replied:
arg no, fixed all these now

ijrsvt commented:
Test failures are unrelated (`test_serve_deploy.py`).

@DmitriGekhtman are you okay if I merge?

DmitriGekhtman commented:
> Test failures are unrelated (`test_serve_deploy.py`).
> 
> @DmitriGekhtman are you okay if I merge?

Looks good, yes!
